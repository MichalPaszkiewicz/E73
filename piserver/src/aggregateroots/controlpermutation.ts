export class ControlPermutation {

    constructor(
        public forwardOn: boolean, 
        public backwardOn: boolean, 
        public leftOn: boolean, 
        public rightOn: boolean, 
        public leftMotorState: number, 
        public rightMotorState: number) {

    }

    isEqual(forwardOn: boolean, backwardOn: boolean, leftOn: boolean, rightOn: boolean) {
        return this.forwardOn == forwardOn &&
            this.backwardOn == backwardOn &&
            this.leftOn == leftOn &&
            this.rightOn == rightOn;
    }

    toString() {
        var self = this;
        return (`     ${self.forwardOn ? 1 : 0}
  ${self.leftOn ? 1 : 0}     ${self.rightOn ? 1 : 0}
     ${self.backwardOn ? 1 : 0}`);
    }
}