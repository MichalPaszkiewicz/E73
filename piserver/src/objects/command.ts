export class Command{
    constructor(public url: string, public response: (data?: any) => void){

    }
}