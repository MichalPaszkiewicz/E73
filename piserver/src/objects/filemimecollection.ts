export class FileMime{
    constructor(public extension: string, public mimeType: string){

    }
    matchesExtension(test: string){
        return this.extension.replace(/./g, "").toLowerCase() == test.replace(/./g, "").toLowerCase();
    }
}

export class FileMimeCollection{
    constructor(public fileMimes: FileMime[]){

    }
    isInCollection(extension: string): boolean{
        return this.fileMimes.some(fm => fm.matchesExtension(extension));
    }
    isARegisteredFileType(fileName: string): boolean{
        var dotIndex = fileName.lastIndexOf(".");
        if(dotIndex == -1){
            return false;
        }
        var extension = fileName.substr(dotIndex);
        return this.isInCollection(extension);
    }
    getFileMime(fileName: string){
        var extension = fileName.substr(fileName.lastIndexOf("."));
        for(var i = 0; i < this.fileMimes.length; i++){
            if(this.fileMimes[i].matchesExtension(extension)){
                return this.fileMimes[i];
            }
        }
    }
}