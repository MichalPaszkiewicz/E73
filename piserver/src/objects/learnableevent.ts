export class LearnableEvent{
    constructor(public action: () => void, public waitTime: number){

    }
} 