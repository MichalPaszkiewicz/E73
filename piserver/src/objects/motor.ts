"use strict"

import {PinState} from "../enums/pinstate";
import {IAmAPin} from "../interfaces/iamapin";
import {IAmAPinFactory} from "../interfaces/iamapinfactory";
import {IAmAMotor} from "../interfaces/iamamotor";

export class Motor<T extends IAmAPin>implements IAmAMotor{
	private _pulseWidth: number = 5;
	private _pulses: number = 100;
	private _currentPulse: number = 0;
	private _speed: number = 0;
	public get speed(){return this._speed}
	private _enablePin: IAmAPin;
	public get enablePin(){ return this._enablePin; }
	private _forwardPin: IAmAPin;
	public get forwardPin(){return this._forwardPin;}
	private _backwardPin: IAmAPin;
	public get backwardPin(){return this._backwardPin;}

	constructor(enablePin: number, forwardPin: number, backwardPin: number, pinFactory: IAmAPinFactory<T>) {
		this._enablePin = pinFactory(enablePin, PinState.OUT);
		this._forwardPin = pinFactory(forwardPin, PinState.OUT);
		this._backwardPin = pinFactory(backwardPin, PinState.OUT);
		this._loop();
	}

	private _loop() {
		var self = this;

		if (self._speed > self._currentPulse / self._pulses) {
			self._enablePin.writeSync(1);
		}
		else {
			self._enablePin.writeSync(0);
		}

		self._currentPulse++;
		self._currentPulse = self._currentPulse % self._pulses;

		setTimeout(() => {
			self._loop();
		}, self._pulseWidth)
	}

	forward() {
		this._speed = 1;
		this._forwardPin.writeSync(1);
		this._backwardPin.writeSync(0);
	}

	backward() {
		this._speed = 1;
		this._forwardPin.writeSync(0);
		this._backwardPin.writeSync(1);
	}

	off() {
		this._speed = 0;
		this._forwardPin.writeSync(0);
		this._backwardPin.writeSync(0);
	}

	setSpeed(newSpeed) {
		if (newSpeed > 1 || newSpeed < 0) {
			throw new Error("The speed must be between 0 and 1");
		}
		this._speed = newSpeed;
	}
}