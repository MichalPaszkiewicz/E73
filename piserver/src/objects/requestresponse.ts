export class RequestResponse{
    constructor(public url: string, public response: (data?: any) => void){

    }
}