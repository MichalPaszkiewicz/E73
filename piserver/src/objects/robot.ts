import {ISteerThings} from "../interfaces/isteerthings";

export class Robot {
    
    private _drive: ISteerThings;

    constructor(drive: ISteerThings){
        this._drive = drive;
    }

    getCommands(){
        return this._drive.getCommands();
    }

    off(){
        this._drive.off();
    }
}