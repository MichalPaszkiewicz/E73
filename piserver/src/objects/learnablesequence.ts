import {LearnableEvent} from "./learnableevent";

export class LearnableSequence{

    constructor(public name: string, public events: LearnableEvent[]){

    }

}