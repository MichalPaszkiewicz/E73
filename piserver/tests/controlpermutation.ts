import {ControlPermutation} from "../src/objects/controlpermutation";

test("toString method returns correct format for forward", () => {

    var controlPermutation = new ControlPermutation(true, false, false, false, 1, 1);

    expect(controlPermutation.toString()).toMatchSnapshot();
});

test("toString method returns correct format for backward", () => {

    var controlPermutation = new ControlPermutation(false, true, false, false, 1, 1);

    expect(controlPermutation.toString()).toMatchSnapshot();
});

test("toString method returns correct format for left", () => {

    var controlPermutation = new ControlPermutation(false, false, true, false, 1, 1);

    expect(controlPermutation.toString()).toMatchSnapshot();
});

test("toString method returns correct format for right", () => {

    var controlPermutation = new ControlPermutation(false, false, false, true, 1, 1);

    expect(controlPermutation.toString()).toMatchSnapshot();
});

test("toString method returns correct format for forward left", () => {

    var controlPermutation = new ControlPermutation(true, false, true, false, 1, 1);

    expect(controlPermutation.toString()).toMatchSnapshot();
});

test("toString method returns correct format for forward right", () => {

    var controlPermutation = new ControlPermutation(true, false, false, true, 1, 1);

    expect(controlPermutation.toString()).toMatchSnapshot();
});

test("toString method returns correct format for backward left", () => {

    var controlPermutation = new ControlPermutation(false, true, true, false, 1, 1);

    expect(controlPermutation.toString()).toMatchSnapshot();
});

test("toString method returns correct format for backward right", () => {

    var controlPermutation = new ControlPermutation(false, true, false, true, 1, 1);

    expect(controlPermutation.toString()).toMatchSnapshot();
});

test("is equal tests a control permutations is equal to directional values", () => {

    var controlPermutation = new ControlPermutation(true, true, false, true, 1, 1);

    expect(controlPermutation.isEqual(true, true, false, true)).toBe(true);
});

test("is equal tests a control permutation is not equal to bad directional values", () => {

    var controlPermutation = new ControlPermutation(false, true, false, false, 1, 1);

    expect(controlPermutation.isEqual(true, false, false, false)).toBe(false);
});