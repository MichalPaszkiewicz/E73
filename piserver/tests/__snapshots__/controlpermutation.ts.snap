// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toString method returns correct format for backward 1`] = `
"     0
  0     0
     1"
`;

exports[`toString method returns correct format for backward left 1`] = `
"     0
  1     0
     1"
`;

exports[`toString method returns correct format for backward right 1`] = `
"     0
  0     1
     1"
`;

exports[`toString method returns correct format for forward 1`] = `
"     1
  0     0
     0"
`;

exports[`toString method returns correct format for forward left 1`] = `
"     1
  1     0
     0"
`;

exports[`toString method returns correct format for forward right 1`] = `
"     1
  0     1
     0"
`;

exports[`toString method returns correct format for left 1`] = `
"     0
  1     0
     0"
`;

exports[`toString method returns correct format for right 1`] = `
"     0
  0     1
     0"
`;
